ocaml/tests/test_repository_helpers.ml

```diff
@@ -741,7 +741,11 @@ module UpdateInfoMetaDataOfXml = Generic.MakeStateless (struct
   end
 
   let transform input =
-    try Ok (UpdateInfoMetaData.of_xml (Xml.parse_string input))
+    try
+      Ok
+        (UpdateInfoMetaData.of_xml (Xml.parse_string input) "updateinfo"
+        |> UpdateInfoMetaData.assert_valid
+        )
```

Comment:
I think the validation should happens in the conversion function, as well as picking the field, simplifying this code into: 

```suggestion
      Ok (UpdateMetaData.of_xml (Xml.parse_string input))
```
This means `UpdateInfoMetadata.of_xml` could be defined as
```ocaml
  let of_xml xml =
    let update_info = Repometadata.of_xml xml "updateinfo" in
    UpdateInfoMetaData.assert_valid update_info ;
    update_info
```
This would make this function much easier to use and less error-prone

Reply:
The `of_xml` will be called in `of_xml_file`. So are you suggesting the `of_xml_file` accepts a function as a parameter? It would look like:
```
module RepoMetaData = struct
  type t = {checksum: string; location: string}

  let default = {checksum= ""; location= ""}

  let of_xml xml data_type =
    ...

  let of_xml_file xml_path of_xml =
    match Sys.file_exists xml_path with
    | false ->
        error "No repomd.xml found: %s" xml_path ;
        raise Api_errors.(Server_error (invalid_repomd_xml, []))
    | true -> (
      match Xml.parse_file xml_path with
      | xml ->
          of_xml xml
      | exception e ->
          error "Failed to parse repomd.xml: %s" (ExnHelper.string_of_exn e) ;
          raise Api_errors.(Server_error (invalid_repomd_xml, []))
    )
end

module UpdateInfoMetaData = struct
  let assert_valid = function
    | RepoMetaData.{checksum= ""; _} | RepoMetaData.{location= ""; _} ->
        error
          "Can't find valid 'checksum' or 'location' in updateinfo in \
           repomod.xml" ;
        raise Api_errors.(Server_error (invalid_repomd_xml, []))
    | md ->
        md

  let of_xml xml =
    RepoMetadata.of_xml "updateinfo" xml |> assert_valid

  let of_xml_file xml_path =
    RepoMetaData.of_xml_file xml_path of_xml
end
```

Reply:
I don't think that change is necessary, similar to how I suggested to implement `UpdateInfoMetaData.of_xml`, `of_xml_path` can be implemented as

```ocaml
  let of_xml_file xml_path =
    let update_info = Repometadata.of_xml_file xml_path "updateinfo" in
    UpdateInfoMetaData.assert_valid update_info ;
    update_info
```